hermes-frontend/src/main/java/pl/allegro/tech/hermes/frontend/publishing/MessageReader.java

```diff
@@ -1,95 +1,57 @@
 
-import javax.servlet.ReadListener;
-import javax.servlet.ServletInputStream;
-import javax.servlet.http.HttpServletRequest;
 import java.io.ByteArrayOutputStream;
 import java.io.IOException;
 import java.util.function.Consumer;
 
 import static pl.allegro.tech.hermes.frontend.publishing.ContentLengthChecker.checkContentLength;
-import static pl.allegro.tech.hermes.frontend.publishing.message.MessageState.State.PARSED;
-
-public class MessageReader implements ReadListener {
-
-    private final ByteArrayOutputStream messageContent = new ByteArrayOutputStream();
-    private final ServletInputStream inputStream;
-    private final Integer chunkSize;
-    private final TopicName topicName;
-    private final HermesMetrics hermesMetrics;
-    private Timer.Context parsingTimerPerTopic;
-    private Timer.Context parsingTimer;
-    private HttpServletRequest request;
-    private MessageState messageState;
-    private Consumer<byte []> onRead;
-    private Consumer<IllegalStateException> onValidationError;
-    private Consumer<Throwable> onOtherError;
-
-    public MessageReader(
-            HttpServletRequest request,
-            Integer chunkSize,
-            TopicName topicName,
-            HermesMetrics hermesMetrics,
-            MessageState messageState,
-            Consumer<byte[]> onRead,
-            Consumer<IllegalStateException> onValidationError,
-            Consumer<Throwable> onOtherError) throws IOException {
-
-        this.request = request;
-        this.messageState = messageState;
-        this.onValidationError = onValidationError;
-        this.onOtherError = onOtherError;
-        this.inputStream = request.getInputStream();
-        this.chunkSize = chunkSize;
-        this.topicName = topicName;
-        this.hermesMetrics = hermesMetrics;
-        this.onRead = onRead;
-
-        messageState.setState(MessageState.State.PARSING);
-        inputStream.setReadListener(this);
-        initParsingTimers();
+import static pl.allegro.tech.hermes.frontend.publishing.message.MessageState.State.FULLY_READ;
+import static pl.allegro.tech.hermes.frontend.publishing.message.MessageState.State.READING;
+import static pl.allegro.tech.hermes.frontend.publishing.message.MessageState.State.READING_TIMEOUT;
+
+class MessageReader {
+
+    public void onMessageRead(HttpServerExchange exchange, MessageState messageState, TopicWithMetrics topicWithMetrics,
+                              MessageTimeout messageTimeout, Consumer<byte[]> messageConsumer, Consumer<IOException> exceptionConsumer) {
+
+        ByteArrayOutputStream messageContent = new ByteArrayOutputStream();
+
+        StartedTimersPair startedTimersPair = topicWithMetrics.startRequestReadTimers();
+
+        messageTimeout.onTimeout((Void) -> {
+            startedTimersPair.close();
+            exchange.getRequestReceiver().pause();
+        });
+
+        exchange.getRequestReceiver().receivePartialBytes(
+                (exchange1, message, last) -> {
```

Comment:
My point is that this lambda can be interpreted as two different callbacks: it matches signature of both `PartialBytesCallback` and `PartialStringCallback` which can be passed to `receivePartialBytes`. I'm not sure if this will not cause some issues one day after jvm update. Maybe we can add explicit argument types to make it clear which callback we want...